server/controllers/userController.js

```diff
@@ -0,0 +1,163 @@
+   * @memberof UsersController
+   */
+  static async signUp(req, res, next) {
+    try {
+      const data = req.body;
+      const { password } = data;
+      data.password = await encryptPassword(password);
+      const user = await userModel.create(data);
+      const token = await generateToken(user);
+
+      return res.status(201).json({
+        status: 'success',
+        data: { token }
+      });
+    } catch (err) {
+      return next(err);
+    }
+  }
+
+  /**
+   *
+   * Login Middleware - Logs user into the application
+   * @static
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @returns {object} Object containing token to the user
+   * @memberof UsersController
+   */
+  static async login(req, res, next) {
+    try {
+      const data = req.body;
+      const { userLogin, password } = data;
+      // Login with username or email address
+      const user = await userModel.findUserInput(userLogin);
+      if (!user) {
+        return res.status(404).json({
+          status: 'fail',
+          data: { message: 'Invalid Login Details' }
+        });
+      }
+      const passwordValid = await decryptPassword(password, user.password);
+      if (!passwordValid) {
+        return res.status(401).json({
+          status: 'fail',
+          data: { message: 'Invalid Login Details' }
+        });
+      }
+      const token = await generateToken(user);
+      return res.status(200).json({
+        status: 'success',
+        data: { token }
+      });
+    } catch (err) {
+      return next(err);
+    }
+  }
+
+  /**
+   *
+   * Forgot Password Middleware - Enables users request new password
+   * @static
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @returns {object} Object containing token to the user
+   * @memberof UsersController
+   */
+  static async forgotPassword(req, res, next) {
+    try {
+      const { userLogin } = req.body;
+      const user = await userModel.findUserInput(userLogin);
+      if (!user) {
+        return res.status(404).json({
+          status: 'fail',
+          data: { message: 'Invalid Login Details' }
+        });
+      }
+      const token = await crypto.randomBytes(20).toString('hex');
+      const mailSubject = 'Reset your password';
+      const mailContent = `You are receiving this mail because you requested
+       for a password change. Please Click on the link below or paste it
+       in a browser to complete the process <br />
+        http://${req.headers.host}/reset/${token} <br/>
+        Link expires in 1 hour time. <br />
+       If you did not request this, please Ignore this email and your
+       password will remain unchanged.
+      `;
+      await emailModule.sendEmailToUser(userLogin, mailSubject, mailContent);
+      await userModel.updateToken(token, userLogin);
+      await userModel.updatePasswordExpires(Date.now() + 3600000, userLogin);
```

Comment:
why does the password expire here?

Reply:
Thank you for pointing this out. It's due to poor naming. The method has nothing to do with password expiration. What the method does is actually to check if the token used in resetting the password has expired. I'll ensure I change the name to suit what the method actually does.